content/articles/fcm-with-spring-and-angular/index.md

```diff
@@ -0,0 +1,500 @@
+                                               .build()
+                               ).build()
+               ).setTopic(notification.topic)
+               .build()
+
+       FirebaseMessaging.getInstance().sendAsync(message)
+
+   }
+}
+```
+As you can see, creating the notifications are pretty straightforward if you know the [builder design pattern](https://howtodoinjava.com/design-patterns/creational/builder-pattern-in-java/). Since our platform is the web, we pass a `WebpushConfig` object to the builder. That object contains the notification data to send. Note that Firebase also gives you the option to set a configuration for Android or IOS devices in a similar way.
+
+Finally, after setting the notification data, we set the topic or token to specify who the message is for. From there, we call the `sendAsync` method to send the message.
+
+To finish this service, let’s create a function to subscribe a user to a specified topic. Before that, let’s create a model object to represent a subscription request:
+
+```kotlin
+// The subscriber field specifies the token of the subscribing user
+data class SubscriptionRequest(val subscriber: String, val topic: String)
+```
+With that class created, here is the function to subscribe a user to a topic:
+
+```kotlin
+fun subscribeToTopic(subscription: SubscriptionRequest){
+
+   FirebaseMessaging.getInstance().subscribeToTopic(listOf(subscription.subscriber), subscription.topic)
+
+}
+```
+
+### Exposing our service through a REST controller
+
+#### Creating the REST Controller
+
+Now that we have our Firebase Cloud Messaging service, all we need to do is create a REST controller to expose it. For sake of example, we are going to have the client send requests to this controller to send notifications to themself. A real production application would probably not be built like that. Assuming you have a solid background in Spring MVC, this should be straightforward:
+
+```kotlin
+package me.john.amiscaray.controllers
+
+import me.john.amiscaray.dtos.SubscriptionRequest
+import me.john.amiscaray.dtos.DirectNotification
+import me.john.amiscaray.dtos.TopicNotification
+import me.john.amiscaray.services.FCMService
+import org.springframework.web.bind.annotation.PostMapping
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RestController
+
+@RestController
+class NotificationController(private val fcm: FCMService) {
+
+   @PostMapping("/notification")
+   fun sendTargetedNotification(@RequestBody notification: DirectNotification){
+
+       fcm.sendNotificationToTarget(notification)
+
+   }
+
+
+   @PostMapping("/topic/notification")
+   fun sendNotificationToTopic(@RequestBody notification: TopicNotification){
+
+       fcm.sendNotificationToTopic(notification)
+
+   }
+
+   @PostMapping("/topic/subscription")
+   fun subscribeToTopic(@RequestBody subscription: SubscriptionRequest){
+
+       fcm.subscribeToTopic(subscription)
+
+   }
```

Comment:
Though, this code formatting looks nice and clear to read.

Sometimes, these extra new lines look too weird after publishing it. They would occupy a majority of the page. Thus, leading the reader to scroll, to read it.

Remove the extra new lines